File: scripts/runtime-payload-manifest.py
#!/usr/bin/env python3
"""Generate or verify the deterministic M4 runtime payload manifest."""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import stat
import sys
from pathlib import Path, PurePosixPath

SCHEMA_VERSION = 1
MANIFEST_NAME = "manifest-v1.json"
REQUIRED_FILES = {
    "Components.noindex/VibeStick Bridge.app/Contents/MacOS/VibeStickBridge",
    "Components.noindex/VibeStick HUD.app/Contents/MacOS/VibeStickHUD",
    "Components.noindex/VibeStick Paste.app/Contents/MacOS/VibeStickPaste",
    "Components.noindex/VibeStick Paste.app/Contents/Resources/VibeStickPaste.build",
    "runtime/bridge/pyproject.toml",
    "runtime/bridge/src/vibe_stick/__main__.py",
}


def fail(message: str) -> "NoReturn":
    raise SystemExit(f"runtime payload: {message}")


def safe_relative_path(value: str) -> bool:
    path = PurePosixPath(value)
    return (
        bool(value)
        and not value.startswith(("/", "~"))
        and "\\" not in value
        and all(part not in ("", ".", "..") for part in path.parts)
    )


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        while chunk := handle.read(1024 * 1024):
            digest.update(chunk)
    return digest.hexdigest()


def payload_files(root: Path) -> list[Path]:
    files: list[Path] = []
    for directory, directory_names, file_names in os.walk(root, followlinks=False):
        directory_path = Path(directory)
        for name in sorted(directory_names):
            candidate = directory_path / name
            if candidate.is_symlink():
                fail(f"symlink is not allowed: {candidate.relative_to(root).as_posix()}")
        for name in sorted(file_names):
            candidate = directory_path / name
            relative = candidate.relative_to(root).as_posix()
            if relative == MANIFEST_NAME:
                continue
            if candidate.is_symlink() or not candidate.is_file():
                fail(f"non-regular file is not allowed: {relative}")
            files.append(candidate)
    return sorted(files, key=lambda item: item.relative_to(root).as_posix())


def entry(root: Path, path: Path) -> dict[str, object]:
    relative = path.relative_to(root).as_posix()
    if not safe_relative_path(relative):
        fail(f"unsafe path: {relative}")
    metadata = path.stat()
    return {
        "mode": stat.S_IMODE(metadata.st_mode),
        "path": relative,
        "sha256": sha256(path),
        "size": metadata.st_size,
    }


def generate(root: Path, version: str) -> None:
    if not root.is_dir():
        fail(f"payload root does not exist: {root}")
    entries = [entry(root, path) for path in payload_files(root)]
    actual = {str(item["path"]) for item in entries}
    missing = sorted(REQUIRED_FILES - actual)
    if missing:
        fail(f"required files are missing: {', '.join(missing)}")
    manifest = {
        "files": entries,
        "payloadVersion": version,
        "schemaVersion": SCHEMA_VERSION,
    }
    destination = root / MANIFEST_NAME
    destination.write_text(
        json.dumps(manifest, ensure_ascii=False, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )
    os.chmod(destination, 0o644)


def verify(root: Path) -> None:
    manifest_path = root / MANIFEST_NAME
    try:
        manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        fail(f"cannot read {MANIFEST_NAME}: {error}")
    if manifest.get("schemaVersion") != SCHEMA_VERSION:
        fail("unsupported schema version")
    if not isinstance(manifest.get("payloadVersion"), str) or not manifest["payloadVersion"].strip():
        fail("payloadVersion is empty")
    declared_entries = manifest.get("files")
    if not isinstance(declared_entries, list):
        fail("files is not an array")

    declared: dict[str, dict[str, object]] = {}
    for raw in declared_entries:
        if not isinstance(raw, dict) or not isinstance(raw.get("path"), str):
            fail("manifest contains an invalid file entry")
        relative = raw["path"]
        if not safe_relative_path(relative):
            fail(f"unsafe path: {relative}")
        if relative in declared:
            fail(f"duplicate path: {relative}")
        declared[relative] = raw

    actual_paths = payload_files(root)
    actual = {path.relative_to(root).as_posix(): path for path in actual_paths}
    if set(actual) != set(declared):
        missing = sorted(set(declared) - set(actual))
        extra = sorted(set(actual) - set(declared))
        fail(f"file set mismatch; missing={missing}, extra={extra}")
    required_missing = sorted(REQUIRED_FILES - set(declared))
    if required_missing:
        fail(f"required files are missing: {', '.join(required_missing)}")

    for relative, path in actual.items():
        expected = declared[relative]
        metadata = path.stat()
        if expected.get("size") != metadata.st_size:
            fail(f"size mismatch: {relative}")
        if expected.get("mode") != stat.S_IMODE(metadata.st_mode):
            fail(f"mode mismatch: {relative}")
        if expected.get("sha256") != sha256(path):
            fail(f"SHA-256 mismatch: {relative}")


def main() -> int:
    parser = argparse.ArgumentParser()
    subparsers = parser.add_subparsers(dest="command", required=True)
    generate_parser = subparsers.add_parser("generate")
    generate_parser.add_argument("root", type=Path)
    generate_parser.add_argument("version")
    verify_parser = subparsers.add_parser("verify")
    verify_parser.add_argument("root", type=Path)
    arguments = parser.parse_args()

    if arguments.command == "generate":
        generate(arguments.root.resolve(), arguments.version)
    else:
        verify(arguments.root.resolve())
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/runtime-payload-manifest.swift
#!/usr/bin/env swift

import CryptoKit
import Darwin
import Foundation

private let schemaVersion = 1
private let manifestName = "manifest-v1.json"
private let requiredFiles: Set<String> = [
    "Components.noindex/VibeStick Bridge.app/Contents/MacOS/VibeStickBridge",
    "Components.noindex/VibeStick HUD.app/Contents/MacOS/VibeStickHUD",
    "Components.noindex/VibeStick Paste.app/Contents/MacOS/VibeStickPaste",
    "Components.noindex/VibeStick Paste.app/Contents/Resources/VibeStickPaste.build",
]

private struct PayloadFile: Codable {
    let mode: UInt16
    let path: String
    let sha256: String
    let size: UInt64
}

private struct PayloadManifest: Codable {
    let files: [PayloadFile]
    let payloadVersion: String
    let schemaVersion: Int
}

private struct ManifestFailure: LocalizedError {
    let message: String
    var errorDescription: String? { "runtime payload: \(message)" }
}

private func fail(_ message: String) throws -> Never {
    throw ManifestFailure(message: message)
}

private func safeRelativePath(_ path: String) -> Bool {
    guard !path.isEmpty,
          !path.hasPrefix("/"),
          !path.hasPrefix("~"),
          !path.contains("\\") else { return false }
    return path.split(separator: "/", omittingEmptySubsequences: false)
        .allSatisfy { !$0.isEmpty && $0 != "." && $0 != ".." }
}

private func payloadFiles(root: URL) throws -> [(path: String, url: URL)] {
    let fileManager = FileManager.default
    guard let enumerator = fileManager.enumerator(
        at: root,
        includingPropertiesForKeys: [.isRegularFileKey, .isSymbolicLinkKey],
        options: []
    ) else { try fail("payload root does not exist: \(root.path)") }

    let rootPrefix = root.standardizedFileURL.path + "/"
    var result: [(String, URL)] = []
    while let url = enumerator.nextObject() as? URL {
        let values = try url.resourceValues(forKeys: [.isRegularFileKey, .isSymbolicLinkKey])
        let standardized = url.standardizedFileURL.path
        guard standardized.hasPrefix(rootPrefix) else { try fail("unsafe enumerated path") }
        let relative = String(standardized.dropFirst(rootPrefix.count))
        if values.isSymbolicLink == true { try fail("symlink is not allowed: \(relative)") }
        guard relative != manifestName else {
            guard values.isRegularFile == true else { try fail("non-regular file is not allowed: \(relative)") }
            continue
        }
        if values.isRegularFile == true { result.append((relative, url)) }
    }
    return result.sorted { $0.0 < $1.0 }
}

private func sha256(_ url: URL) throws -> String {
    let handle = try FileHandle(forReadingFrom: url)
    defer { try? handle.close() }
    var digest = SHA256()
    while true {
        let data = try handle.read(upToCount: 1_048_576) ?? Data()
        if data.isEmpty { break }
        digest.update(data: data)
    }
    return digest.finalize().map { String(format: "%02x", $0) }.joined()
}

private func entry(relative: String, url: URL) throws -> PayloadFile {
    guard safeRelativePath(relative) else { try fail("unsafe path: \(relative)") }
    let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
    guard let size = (attributes[.size] as? NSNumber)?.uint64Value,
          let mode = (attributes[.posixPermissions] as? NSNumber)?.uint16Value else {
        try fail("cannot read metadata: \(relative)")
    }
    return PayloadFile(mode: mode, path: relative, sha256: try sha256(url), size: size)
}

private func generate(root: URL, version: String) throws {
    var isDirectory: ObjCBool = false
    guard FileManager.default.fileExists(atPath: root.path, isDirectory: &isDirectory), isDirectory.boolValue else {
        try fail("payload root does not exist: \(root.path)")
    }
    guard !version.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        try fail("payloadVersion is empty")
    }
    let entries = try payloadFiles(root: root).map { try entry(relative: $0.path, url: $0.url) }
    let actual = Set(entries.map(\.path))
    let missing = requiredFiles.subtracting(actual).sorted()
    guard missing.isEmpty else { try fail("required files are missing: \(missing.joined(separator: ", "))") }

    let encoder = JSONEncoder()
    encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
    var data = try encoder.encode(PayloadManifest(
        files: entries,
        payloadVersion: version,
        schemaVersion: schemaVersion
    ))
    data.append(0x0A)
    let destination = root.appendingPathComponent(manifestName)
    try data.write(to: destination, options: .atomic)
    try FileManager.default.setAttributes([.posixPermissions: 0o644], ofItemAtPath: destination.path)
}

private func verify(root: URL) throws {
    let manifestURL = root.appendingPathComponent(manifestName)
    let values = try manifestURL.resourceValues(forKeys: [.isRegularFileKey, .isSymbolicLinkKey])
    guard values.isRegularFile == true, values.isSymbolicLink != true else {
        try fail("cannot read \(manifestName)")
    }
    let manifest: PayloadManifest
    do { manifest = try JSONDecoder().decode(PayloadManifest.self, from: Data(contentsOf: manifestURL)) }
    catch { try fail("cannot read \(manifestName): \(error.localizedDescription)") }
    guard manifest.schemaVersion == schemaVersion else { try fail("unsupported schema version") }
    guard !manifest.payloadVersion.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        try fail("payloadVersion is empty")
    }

    var declared: [String: PayloadFile] = [:]
    for item in manifest.files {
        guard safeRelativePath(item.path) else { try fail("unsafe path: \(item.path)") }
        guard declared[item.path] == nil else { try fail("duplicate path: \(item.path)") }
        declared[item.path] = item
    }
    let actualFiles = try payloadFiles(root: root)
    let actual = Dictionary(uniqueKeysWithValues: actualFiles.map { ($0.path, $0.url) })
    guard Set(actual.keys) == Set(declared.keys) else {
        let missing = Set(declared.keys).subtracting(actual.keys).sorted()
        let extra = Set(actual.keys).subtracting(declared.keys).sorted()
        try fail("file set mismatch; missing=\(missing), extra=\(extra)")
    }
    let requiredMissing = requiredFiles.subtracting(declared.keys).sorted()
    guard requiredMissing.isEmpty else {
        try fail("required files are missing: \(requiredMissing.joined(separator: ", "))")
    }
    for relative in actual.keys.sorted() {
        guard let url = actual[relative], let expected = declared[relative] else { continue }
        let observed = try entry(relative: relative, url: url)
        guard observed.size == expected.size else { try fail("size mismatch: \(relative)") }
        guard observed.mode == expected.mode else { try fail("mode mismatch: \(relative)") }
        guard observed.sha256 == expected.sha256 else { try fail("SHA-256 mismatch: \(relative)") }
    }
}

do {
    let arguments = Array(CommandLine.arguments.dropFirst())
    guard arguments.count == 2 || arguments.count == 3 else {
        try fail("usage: runtime-payload-manifest.swift generate ROOT VERSION | verify ROOT")
    }
    let root = URL(fileURLWithPath: arguments[1], isDirectory: true).standardizedFileURL
    switch arguments[0] {
    case "generate" where arguments.count == 3:
        try generate(root: root, version: arguments[2])
    case "verify" where arguments.count == 2:
        try verify(root: root)
    default:
        try fail("usage: runtime-payload-manifest.swift generate ROOT VERSION | verify ROOT")
    }
} catch {
    let message = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
    FileHandle.standardError.write(Data((message + "\n").utf8))
    exit(1)
}
